### ISSUE TITLE
Attach line & file info to anyhow::Error and anyhow::Context

### ISSUE THREAD

**Issue by vultix** [dtolnay/anyhow#139]
First off, thank you for this wonderful library!

### Context
It's often difficult to find where in your code an error occurred.  Rust's backtraces help with this, but are often so verbose that it's still difficult.

### Proposal

It would be great if you could simply glance at the error output to get file and line information.  This could be implemented using `#[track_caller]` and can be optionally turned on using a feature flag.

I'm not sure what format you'll want, but I'm imagining something like this:
```
Error: Failed to parse file: hello.txt @ main.rs:7:10

Caused by:
    0: Unable to convert this string into a float: '1.5a' @ main.rs:19:24
    1: invalid float literal @ main.rs:25:32
    
Stack backtrace:
...
```

I'm willing to make a PR if you like this feature!

**Comment by mkeedlinger** [comment]
@vultix I would be very interested in this feature!

**Comment by drbartling** [comment]
I would like to see this as a debug feature.  Off by default for release builds.

**Comment by Rua** [comment]
I came here looking for this, so this feature would be very welcome!

**Comment by drbartling** [comment]
For the time being I created this macro for my use:

```rust
#[cfg(feature = "file_lines")]
macro_rules! line_leader {
    () => {
        format!("{}:{}", file!(), line!());
    };
}

#[cfg(feature = "file_lines")]
macro_rules! f_context {
    ($msg:literal $(,)?) => {
        format!("{} {}", line_leader!(), $msg);
    };
    ($fmt:expr, $($arg:tt)*) => (
        format!("{} {}", line_leader!(), format!($fmt, $($arg)*));
    )
}
#[cfg(not(feature = "file_lines"))]
macro_rules! f_context {
    ($msg:literal $(,)?) => {
        $msg
    };
    ($fmt:expr, $($arg:tt)*) => (
        format!($fmt, $($arg)*);
    )
}
```

I use it like this:
```rust
            can_error(&arg)
                .context(f_context!("Did error with {:?}", &arg));
```

**Comment by bruwozniak** [comment]
@dtolnay is the https://github.com/dtolnay/anyhow/tree/location branch still planned to be released? Right now we're trying to extract location via the backtraces and as my colleague kindly put it, it's "gnarly".

**Comment by laralove143** [comment]
> I would like to see this as a debug feature. Off by default for release builds.

i dont agree with this, i think it can be a useful feature for release builds for debugging bugs, i'd rather have this as a separate feature

**Comment by rudolphfroger** [comment]
For me a separate feature would also be a good solution. I'd like to track the location, especially also in release builds because it increases the possibility to understand what happened.

**Comment by ajantti** [comment]
As someone coming from C++ where I implemented almost exactly the same thing as Result + anyhow, this gets my vote. In my library I came to depend on it so much it was the immediate first thought for me when I found anyhow. I just implemented my own macro for my library, but I liked the formatting in [drbartling](https://github.com/drbartling)'s answer so I continued it a bit:
```rust
#[cfg(feature = "file_lines")]
macro_rules! file_line_leader {
    () => {
        format!("{}:{}", file!(), line!());
    };
}

#[cfg(feature = "file_lines")]
macro_rules! atry {
    ($x:expr) => {
        anyhow::Result::from($x).context(file_line_leader!())?
    };
    ($x:expr, $msg:literal $(,)?) => {
        anyhow::Result::from($x).context(format!("{}: {}", file_line_leader!(), $msg))?
    };
    ($x:expr, $fmt:expr, $($arg:tt)*) => {
        anyhow::Result::from($x).context(format!("{}: {}", file_line_leader!(), format!($fmt, $($arg)*)))?
    };
}
#[cfg(not(feature = "file_lines"))]
macro_rules! atry {
    ($x:expr) => {
        anyhow::Result::from($x)?
    };
    ($x:expr, $msg:literal $(,)?) => {
        anyhow::Result::from($x).context($msg)?
    };
    ($x:expr, $fmt:expr, $($arg:tt)*) => {
        anyhow::Result::from($x).context(format!($fmt, $($arg)*))?
    };
}
```
So it can be used like this:
```rust
let username = atry!(read_username_from_file(), "Couldn't get from file {}", filename);
```
The question is though, should the macro include the ? or not? Usually after checking for error, you want the value so I think for majority of the cases, you'd want the ? anyway, but if you want to return the value, then you have to rewrap the result. And adding two different macros would be just silly. Well, except in my C++ one  I had another macro that allows for changing the error kind, but I'll leave this as such for now.

**Comment by taewoo-prestolabs** [comment]
Any update for this track_caller feature?

**Comment by codingskynet** [comment]
Any update for this track_caller(ocation) feature?

**Comment by cybersoulK** [comment]
Having the same issue where it's very hard to know where the error is originating, there are similar methods called with the same error, and adding manual context to each is not a great option.

**Comment by AaronKutch** [comment]
I just updated my crate https://crates.io/crates/stacked_errors and it shows how you can attach `#[track_caller]` location with errors to make a programmatic backtrace that is entirely independent of the `RUST_BACKTRACE` system. I'm coming from the opposite direction where now I might want to add standard backtraces in parallel with this. The only semi-ugly thing is that you need to call `.stack()` at any point where you are propogating an error and want it to log at that point with the programmatic backtraces. It might be possible to automatically do this with the `Try` trait (std might need to add `#[track_caller]` on their side).
